src/Platforms/devices.h

```diff
@@ -155,30 +172,34 @@ inline void set_appropriate_device_num(int num)
   int deviceCount;
   cudaGetDeviceCount(&deviceCount);
   int num_appropriate=0, device=0;
-  bool set_cuda_device=false;
+  bool set_cuda_device=true;
   for (device = 0; device < deviceCount; ++device)
   {
     cudaDeviceProp deviceProp;
     cudaGetDeviceProperties(&deviceProp, device);
     if (((deviceProp.major >= 1) && (deviceProp.minor >= 3)) ||
         deviceProp.major >= 2)
     {
+      if (num_appropriate<gpu::device_group_size)
+        gpu::device_group_numbers[num_appropriate] = device;
```

Comment:
Is it possible to get ride of device_group_numbers?
get_num_appropriate_devices() already check device properties and if no appropriate device exists, the code aborts in get_device_num().
I think it seems always that device_group_numbers[i]=i

Reply:
If we want to support GPU oversubscription or more interesting MPI layouts the answer is no. Also, while it costs almost no memory (a vector of ints) it simplifies the code downstream quite a bit. Furthermore, It could be used to support the "undersubscription" you asked for (6 MPI ranks sharing 3 GPUs) as a config option in our code. I did think about reading the gpu_sharing flag in as a number to use for that but then realized I can achieve the same thing externally.